woocommerce/woocommerce
# Orders: Add method for obtaining payment info for use in emails and printed receipts

### Changes proposed in this Pull Request:

This adds a public method to Order objects, `get_payment_card_info()`, that returns an associative array of data about the credit card used for the payment in the order. Under the hood there is a new Internal class, `PaymentInfo` which currently only retrieves card info when WooPayments is the gateway for the payment in the order, but other gateways can hook into it and add their own data. The class could be added on to in the future if we have a use case for getting other types of payment info besides credit cards.

The purpose of this new method is so that certain card data can be included in email and printed receipts when it's available. Adding that data to those respective templates will happen in separate PRs.

The wcpay part of this borrows heavily from the work done in the [ReceiptRenderingEngine class](https://github.com/woocommerce/woocommerce/blob/8de9cd3b360a6b533ed70da1141adbd3ba8b8b25/plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php#L420), but is a bit different because when I was testing this with WooPayments, I was not able to get the same data format (`card_present`) that the receipt engine assumes. I'm not sure why this is, but it might have something to do with me testing with web payments rather than mobile app Tap-To-Pay. cc @Konamiman in case you have some insight here. Also, we don't need all the same payment info items for email receipts, only the CC's last 4 digits and the card brand.

Fixes #51823

### How to test the changes in this Pull Request:

This is probably best tested via [WooCommerce Live Branches](https://github.com/woocommerce/woocommerce/blob/8de9cd3b360a6b533ed70da1141adbd3ba8b8b25/plugins/woocommerce-beta-tester/userscripts/README.md#L1-L1), as setting up WooPayments and generating some test payments in a local environment is very challenging.

1. On your test site, install and activate [WooPayments](https://wordpress.org/plugins/woocommerce-payments/), and set it up in sandbox mode.
2. On the front end, go through the process to buy a product and use one of the WooPayments test cards during checkout (there should be a link to the list of test cards on the checkout screen).
3. On the back end, make note of the order ID.
4. Now start a wp-cli session for your test site and enter the shell prompt, `wp shell`.
5. In the shell, use the order ID with the following commands:
    ```
    $order = wc_get_order( {order ID} )
    $order->get_payment_card_info()
    ```
6. You should get an array containing for items, `payment_method`, `brand`, `icon`, and `last4`. Their values should make sense for the test card you used for the order payment.
7. Try making another order with a different payment method and see what `get_payment_card_info()` gives you.
8. Make sure the card icons still work in the receipt engine.
